mixbox/namespaces.py

```diff
@@ -305,41 +298,34 @@ def set_preferred_prefix_for_namespace(self, ns_uri, prefix,
             DuplicatePrefixError: If ``prefix`` already maps to a different
                 namespace.
         """
-
-        ni = self.__ns_uri_map.get(ns_uri)
-        if not ni:
-            raise NamespaceNotFoundError(ns_uri)
+        ni = self.__lookup_uri(ns_uri)
 
         if not prefix:
             ni.preferred_prefix = None
-        else:
-            if add_if_not_exist:
-                self.__check_prefix_conflict(ni, prefix)
-                ni.prefixes.add(prefix)
-                self.__prefix_map[prefix] = ni
-            else:
-                if prefix not in ni.prefixes:
-                    raise PrefixNotFoundError(prefix)
-
+        elif prefix in ni.prefixes:
             ni.preferred_prefix = prefix
+        elif add_if_not_exist:
+            self.add_prefix(ns_uri, prefix, set_as_preferred=True)
+        else:
+            raise PrefixNotFoundError(prefix)
 
     def __merge_schema_locations(self, ni, incoming_schemaloc):
         """Merge incoming_schemaloc into the given `_NamespaceInfo`, ni.  If we
         don't have one yet and the incoming value is non-None, update ours
-        with theirs.  This modifies ni."""
-        if ni.schema_location and incoming_schemaloc:
-            if ni.schema_location != incoming_schemaloc:
-                raise ConflictingSchemaLocationError(ni.uri,
-                                                     ni.schema_location,
-                                                     incoming_schemaloc)
-        elif ni.schema_location is None:
-                ni.schema_location = incoming_schemaloc or None
+        with theirs.  This modifies ni.
+        """
+        if ni.schema_location == incoming_schemaloc:  # TODO (bworrell): empty strings?
```

Comment:
I wasn't sure if we'd run into trouble if we compared an empty string to `None`. If so, we'd need to rethink the if/elif/else chain.